packages/network/src/fetch.ts

```diff
@@ -0,0 +1,42 @@
+import { Request, fetch } from 'cross-fetch';
+
+export type Fetch = typeof fetch;
+
+export type Middleware = (request: Request, next: (request: RequestInfo) => void) => Promise<void>;
+
+export interface BuildFetchOptions {
+  // Whether to force earlier built fetches to error making the most recent //
+  // invokation the authoritive fetch. You will typically only want to set this to
+  // false for testing. Defaults to true.
+  disablePrior?: boolean;
+  // What message to throw if a user tries to invoke a disabled fetch. Useful
+  // to help users know where to import fetch from rather than build it //
+  // themselves.
+  disableMessage?: string;
+  // override the default cross-fetch implementation
+  fetch: Fetch;
+};
+
+export function buildFetch(middleware: Middleware[], options?: BuildFetchOptions): Fetch {
+    const _fetch = options?.fetch || fetch;
+
+    return async (input: RequestInfo, init?: RequestInit) => {
+        const request = new Request(input, init);
+
+        for(const ware of middleware) {
+            await new Promise((resolve, reject) => {
+                try {
+                    ware(request, (request: RequestInfo) => {
+                        request = request;
+    
+                        resolve(request)
+                    });
+                } catch(ex) {
+                    reject(ex);
+                }
+            });
+        }
+        
+        return _fetch(request);
```

Comment:
I find the looping easier to follow, however we've now lost the types.

I think part of the trickiness here is it makes a lot more sense to loop from right to left. I'll push up a commit to demonstrate what I mean; LMKWYT.

Reply:
thanks for showing what you meant in a commit.